README.rst
========
Exordium
========

Exordium is a read-only web-based music library system for Django.
Exordium will read mp3 and ogg files from the host filesystem and provide
an online interface to browse, download (as zipfiles or otherwise), and
stream.

The HTML5 media player `jPlayer <http://jplayer.org/>`_ is used to provide
arbitrary streaming of music.

Exordium was built with a very specific set of operational goals and does
not attempt to be a generic library suitable for widespread use.  There are,
in fact, no configuration options beyond those to define the file paths/URLs
necessary for basic usage.  Patches to add/change functionality will be
happily received so long as they don't interfere with or disable the current
functionality by default, but there is no internal development goal to make
Exordium a generic solution.

Detailed information about what Exordium expects from your library, and its
assumptions and limitations, can be found either in the ``docs/`` directory,
`Exordium's Website <https://apocalyptech.com/exordium>`_, or
`exordium.readthedocs.io <https://exordium.readthedocs.io/>`_.  It's
highly recommended to at least glance through these to get a feel for how
Exordium will function.

The name "Exordium" comes from the fictional technology of the same name in
Alastair Reynolds' "Revelation Space" novels.  It's not a perfect name for
the app, given that the Revelation Space *Exordium* would make a pretty
lousy music library, but at least there's some element of data storage and
retrieval.  Exordium the *web-based music library*, as opposed to its
fictional counterpart, is only capable of retrieving music which has been
imported to it in the past.  I'll be sure to contact all the major news
organizations if I figure out a way to get it to retrieve music stored in the
future.

Requirements
------------

Exordium requires at least Python 3.8 *(tested on 3.9)*, and Django 4.0.
It requires the following additional third-party modules:

- mutagen (built on 1.45)
- Pillow (built on 9.0)
- django-tables2 (built on 2.4)
- django-dynamic-preferences (built on 1.11), which in turn requires:

  - six (built on 1.16.0)
  - persisting-theory (built on 0.2.1)

One unit test module additionally requires django-test-migrations (tested
with 1.2.0), but that's not required to run it.

Getting Exordium
----------------

Exordium is available to install on PyPI via ``pip install django-exordium``.
PyPI also hosts Python packages for Exordium in both source and
`Wheel <https://pypi.python.org/pypi/wheel>`_ formats, at 
https://pypi.python.org/pypi/django-exordium/.  Source and Wheel downloads
of all released versions can also be found at Exordium's hompeage at
https://apocalyptech.com/exordium/.

Exordium sourcecode is hosted at `GitHub <https://github.com/apocalyptech/exordium/>`_,
and sourcecode archives of released versions can be found there at
https://github.com/apocalyptech/exordium/releases

Installation
------------

These instructions assume that you already have a Django project up and
running.

1. Install Exordium via ``pip install django-exordium``

   - If Exordium hasn't been installed via ``pip`` or some other method which
     automatically installs dependencies, install its dependencies::

        pip install -r requirements.txt

2. Add exordium, django_tables2, and dynamic_preferences to your
   ``INSTALLED_APPS`` setting like this::

     INSTALLED_APPS = [
         ...
         'exordium',
         'django_tables2',
         'dynamic_preferences',
         'dynamic_preferences.users.apps.UserPreferencesConfig',
     ]

3. *(Optional)* For jPlayer streaming to work properly on a "live"
   install, the `Cross-Origin-Opener-Policy <https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cross-Origin-Opener-Policy>`_
   HTTP header has to be set properly.  (This will generally not
   be an issue when running Django in "test" mode via ``runserver``.)
   Django defaults to using ``same-origin``, but unless your static
   content delivery also uses the same header, launching the streaming
   window will fail.  You can set the header to ``same-origin-allow-popups``
   inside ``settings.py`` to make this work, or ensure that your static
   files set the proper header.  (Making sure static files use
   ``Cross-Origin-Opener-Policy: same-origin`` just like Django will
   do the trick.)  Setting the Django default can be done with::

    SECURE_CROSS_ORIGIN_OPENER_POLICY = 'same-origin-allow-popups'

   If your static content isn't served from the same protocol/hostname/port
   as Django itself, you will likely have to set either Django or your
   static files' value to ``unsafe-none`` instead.

4. Include the exordium URLconf in your project ``urls.py`` like this::

     path('exordium/', include('exordium.urls')),
   
5. Run ``python manage.py migrate dynamic_preferences`` to create the
   Dynamic Preferences models, if this wasn't already configured on your
   Django install.

6. Run ``python manage.py migrate exordium`` to create the Exordium models.

7. Run ``python manage.py loaddata --app exordium initial_data`` to load
   some initial data into the database.  *(This is not actually strictly
   speaking necessary - the app will create the necessary data
   automatically if it's not found.)*

8. If running this from a "real" webserver, ensure that it's configured
   to serve Django static files. Then run ``python manage.py collectstatic``
   to get Exordium's static files in place.  If you didn't want to set
   ``same-origin-allow-popups`` for Django's COOP header, make sure that
   your server sends a ``Cross-Origin-Opener-Policy: same-origin`` header
   along with these static files, or possibly ``unsafe-none`` if the
   static files protocol/hostname/port doesn't match Django's.

9. Either start the development server with ``python manage.py runserver``
   or bring up your existing server.  Also ensure that you have a webserver
   configured to allow access directly to your music library files, and 
   optionally to the zipfile downloads Exordium will create.
   
10. Visit the administrative area in *Dynamic Preferences > Global preferences*
    and set the values for the following:

    - **Exordium Library Base Path**: This is what defines where your music
      library can be found on disk.
    - **Exordium Media URL for HTML5**: This is the base URL which provides
      direct access to the files in your library, used by the HTML streaming
      player.  Omit the trailing slash, though things will probably work fine
      even if it's in there.  Without this set properly, the streaming
      player will not work properly.  Note that if your base URL for Exordium
      is https, this will have to be https as well, to avoid browser errors.
    - **Exordium Media URL for m3u**: This is the base URL which provides
      direct access to the files in your library, used by the m3u Playlist
      functionality, and also the direct song download links when enumerating
      tracks.  This can be the same as the HTML5 URL.  Omit the slash, though
      things will probably work fine even if it's in there.  Without this set
      properly, m3u playlists and direct track downloads will not work
      properly.  This URL can be http even if the main site is https.
    - **Exordium Zip File Generation Path**: Path on the filesystem to store
      zipfile album downloads.  This is the one location in which the user
      running Django needs write access.
    - **Exordium Zip File Retrieval URL**: This is the base URL providing
      web access to that zipfile directory.  Note that if your base URL for
      Exordium is https, this will have to be https as well, to avoid
      browser errors.

    Without the last two options, Exordium will still function fine, but the
    album-download button will not be rendered.  Exordium will also function
    without the "*Exordium Media URL*" options being set properly, though
    with the caveats mentioned above.

11. If Zipfile downloads are configured, a process should be put into place
    to delete the zipfiles after a period of time.  I personally use a cronjob
    to do this::

      0 2 * * * /usr/bin/find /var/audio/exordiumzips -type f -name "*.zip" -mtime +2 -print -exec unzip -v {} \; -exec rm {} \;

12. Visit the **Library Upkeep** link from the Exordium main page and click on
    "Start Process" to begin the initial import into Exordium!


CHANGELOG.rst
1.4.4 (unreleased)
------------------

*(nothing yet)*

1.4.3 (2023-03-21)
------------------

**Bugfixes/Tweaks**

- Added an extra normalization char (ō / Ō)

1.4.2 (2022-03-15)
------------------

**Bugfixes/Tweaks**

- Fixed some more edge cases with file updates (specifically cases
  where all files in an album get updated, but a subset are also
  renamed -- only the renamed files were actually getting updated
  in the database properly).
- Expanded the internal "normalization" character set to handle some
  more cases, namely super-and-subscript numerals.

1.4.1 (2022-03-04)
------------------

**Bugfixes/Tweaks**

- Added some custom config for readthedocs.org so that our RTD docs
  get generated on that end again.

1.4.0 (2022-03-03)
------------------

**Configuration Updates Needed**

- To continue using jPlayer HTML5 streaming properly on prod installations
  where static content is served via an external webserver, some configuration
  related to `Cross Origin Opener Policy <https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cross-Origin-Opener-Policy>`_
  needs to be tweaked, either in your Django site's ``settings.py``, or
  in the webserver config.  So, one of the following must be done:

  - Set ``SECURE_CROSS_ORIGIN_OPENER_POLICY = 'same-origin-allow-popups'``
    in ``settings.py``, so that the jPlayer popup doesn't need to supply
    its own COOP headers.  Or:
  - Configure your webserver to add ``Cross-Origin-Opener-Policy: same-origin``
    to static file delivery.
  - If your static content isn't served from the same protocol/hostname/port
    as Django itself, you will likely have to set either Django or your
    static files' value to ``unsafe-none`` instead.

- The "Exordium Media URL" preference has been split into two separate
  preferences: one for HTML5 streaming, and one for m3u playlists and
  direct track downloads.

  - If migrations are run prior to accessing the updated site (which
    is always a good idea anyway), the previous "Media URL" value will
    be copied over to both new preferences, which should allow the
    install to continue to work without problems.
  - If Exordium is visited *prior* to running migrations, the two new
    prefs will instead inherit their defaults, as if Django was installed
    for the first time, so they would have to be manually set via the
    web UI in that case.

**Bugfixes/Tweaks**

- Updated for Django 4.x compat, along with more recent versions of our
  other dependencies.
- Split unit tests into multiple files
- Added HTML <label> entries for radio/checkbox text on Library Upkeep page,
  so the labels themselves are clickable instead of just the buttons.
- Link directly to Exordium preferences in Global Preferences, rather than
  just to the main Global Preferences screen.
- Fixed some edge cases with library updates
- Prevent jPlayer popup from opening too many popups when
  `Cross Origin Opener Policy <https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cross-Origin-Opener-Policy>`_
  problems are encountered.
- Added License text to the generated docs.  Not sure why I never included
  those originally!

1.3.4 (2021-03-25)
------------------

**Bugfixes/Tweaks**

- Expanded the internal "normalization" character set to handle a bunch
  more Greek characters specifically, and probably a few others as well.
  This is used for normalizing filenames in zip downloads, and normalizing
  search strings.

1.3.3 (2020-05-28)
------------------

**New Features**

- Added support for Ogg Opus files (``.opus``)

1.3.2 (2018-09-20)
------------------

**Bugfixes/Tweaks**

- Stupid little formatting fix in README and docs requirements RST

1.3.1 (2018-09-20)
------------------

**Bugfixes/Tweaks**

- Use a label for our "live album" checkbox so the text can be clicked
  in addition to the checkbox itself
- Disallow django-tables2 >= 2.0 until an issue with that has been either
  fixed or denied: https://github.com/jieter/django-tables2/issues/621
  (we can work around, if they opt not to merge that fix, but I'd prefer
  to use the fix on that Issue)

1.3.0 (2018-01-02)
------------------

**Bugfixes/Tweaks**

- Updates to work properly with Django 2.0
- Use time-added as secondary sorting for albums when sorting by Year

1.2.1 (2017-11-28)
------------------

**Bugfixes/Tweaks**

- Updates to documentation, to have ``django-dynamic-preferences``
  properly configured.

1.2.0 (2017-11-28)
------------------

**Bugfixes/Tweaks**

- Updates to work properly with Django 1.11 and
  ``django-dynamic-preferences`` >= 1.0
- Fixed live recording checkbox when not logged in to Django

1.1.1 (2016-12-30)
------------------

**Bugfixes/Tweaks**

- Fixed the release date in the Changelog.  Bah.

1.1.0 (2016-12-30)
------------------

**New Features**

- Added support for M4A audio files

**Bugfixes/Tweaks**

- Added a few more "normalization" characters, for easy searching
  from the web UI and correct association across possibly-
  inconsistent tags.  Specifically: İ, ğ, and ş.  Also fixed
  normalizing filenames (for zipfile downloads) for capital Ç.
- Fixed album summary information when some tracks have classical
  music tags (ensemble, composer, conductor) but other tracks
  don't.  (Explicitly say that not all tracks have the tags.)
- Change the display order of a few elements on the album download
  page, and use an HTML ``<meta>`` tag to automatically queue up
  the download, rather than only having the direct link.
- Override table footers to always include item counts, as was
  present in ``django-tables2`` 1.2.6 but patched out in 1.2.7.
- Use newlines when reporting multiple artists in tables, to keep
  the table width down as much as possible.

1.0.3 (2016-11-22)
------------------

**Bugfixes/Tweaks**

- Fixed admin area to allow blank album art, song, and
  artist fields, where the fields shouldn't be required

1.0.2 (2016-10-21)
------------------

**Bugfixes/Tweaks**


exordium/migrations/0001_initial.py
# -*- coding: utf-8 -*-
# Generated by Django 1.10 on 2016-09-30 18:47
from __future__ import unicode_literals

from django.db import migrations, models
import django.db.models.deletion
import django.utils.timezone


class Migration(migrations.Migration):

    initial = True

    dependencies = [
    ]

    operations = [
        migrations.CreateModel(
            name='Album',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('name', models.CharField(max_length=255)),
                ('normname', models.CharField(max_length=255)),
                ('year', models.IntegerField(default=0)),
                ('miscellaneous', models.BooleanField(default=False)),
                ('live', models.BooleanField(default=False)),
                ('time_added', models.DateTimeField(default=django.utils.timezone.now)),
                ('art_filename', models.CharField(default=None, max_length=4096, null=True)),
                ('art_mtime', models.IntegerField(default=0, null=True)),
                ('art_ext', models.CharField(default=None, max_length=4, null=True)),
                ('art_mime', models.CharField(default=None, max_length=64, null=True)),
            ],
        ),
        migrations.CreateModel(
            name='AlbumArt',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('size', models.CharField(choices=[('album', 'Album Listing'), ('list', 'Inside Lists')], max_length=5)),
                ('resolution', models.PositiveSmallIntegerField(default=0)),
                ('from_mtime', models.IntegerField(default=0)),
                ('image', models.BinaryField()),
                ('album', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, to='exordium.Album')),
            ],
        ),
        migrations.CreateModel(
            name='Artist',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('name', models.CharField(max_length=255, unique=True)),
                ('normname', models.CharField(max_length=255, unique=True)),
                ('prefix', models.CharField(max_length=32)),
                ('various', models.BooleanField(default=False)),
            ],
        ),
        migrations.CreateModel(
            name='Song',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('filename', models.CharField(max_length=4096)),
                ('title', models.CharField(max_length=255)),
                ('normtitle', models.CharField(max_length=255)),
                ('year', models.IntegerField()),
                ('tracknum', models.SmallIntegerField(verbose_name='#')),
                ('raw_artist', models.CharField(max_length=255)),
                ('raw_group', models.CharField(default='', max_length=255)),
                ('raw_composer', models.CharField(default='', max_length=255)),
                ('raw_conductor', models.CharField(default='', max_length=255)),
                ('filetype', models.CharField(choices=[('mp3', 'mp3'), ('ogg', 'ogg')], max_length=3)),
                ('bitrate', models.IntegerField()),
                ('mode', models.CharField(choices=[('ABR', 'ABR'), ('CBR', 'CBR'), ('VBR', 'VBR')], max_length=3)),
                ('size', models.IntegerField()),
                ('length', models.IntegerField()),
                ('time_added', models.DateTimeField(default=django.utils.timezone.now)),
                ('time_updated', models.IntegerField(default=0)),
                ('sha256sum', models.CharField(max_length=64)),
                ('album', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, to='exordium.Album')),
                ('artist', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, to='exordium.Artist')),
                ('composer', models.ForeignKey(blank=True, default=None, null=True, on_delete=django.db.models.deletion.CASCADE, related_name='composer_set', to='exordium.Artist')),
                ('conductor', models.ForeignKey(blank=True, default=None, null=True, on_delete=django.db.models.deletion.CASCADE, related_name='conductor_set', to='exordium.Artist')),
                ('group', models.ForeignKey(blank=True, default=None, null=True, on_delete=django.db.models.deletion.CASCADE, related_name='group_set', to='exordium.Artist')),
            ],
        ),
        migrations.AddField(
            model_name='album',
            name='artist',
            field=models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, to='exordium.Artist'),
        ),
        migrations.AlterUniqueTogether(
            name='albumart',
            unique_together=set([('album', 'size')]),
        ),
        migrations.AlterUniqueTogether(
            name='album',
            unique_together=set([('artist', 'name')]),
        ),
    ]
